### PR TITLE
Add considerations for string truncation.

### LINKED ISSUE #973: truncation to 64-byte upper limit doesn't mention character boundaries
https://w3c.github.io/webauthn/#dictionary-pkcredentialentity

When referring to the `name` the spec says:

> Authenticators MUST accept and store a 64-byte minimum length for a name member’s value. Authenticators MAY truncate a name member’s value to a length equal to or greater than 64 bytes.

Note that the specification does not require truncation on a Unicode character boundary. Arbitrary truncation at a 64-byte limit on a multibyte encoding such as UTF-8 can corrupt the last character in the string. The spec should require that the truncation occur on a character boundary (is there a reason you didn't use character count instead of byte count in the first place?)

### PR DESCRIPTION
**Pull request by agl**
Fixes #973.

***
<a href="https://pr-preview.s3.amazonaws.com/agl/webauthn/pull/1205.html" title="Last updated on May 10, 2019, 10:20 PM UTC (f041335)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/w3c/webauthn/1205/5d11e62...agl:f041335.html" title="Last updated on May 10, 2019, 10:20 PM UTC (f041335)">Diff</a>

### CODE REVIEW

**Comment by agl** [comment]
> Although I understand allowing it, I'd suggest saying instead:

Done.

> The comment about grapheme clusters is okay, but I would suggest referring to one of our references.

Happy to take suggestions about what to reference here.

> The text about "understanding string encodings and Unicode character properties" might be better as just "understanding character encodings".

Avoiding truncating a grapheme cluster would require Unicode property tables, I think?

> I'd recommend the use of UTF-8 explicitly. Imposing additional checks, including the check for isolated surrogate code points, seems like overkill.

Webauthn doesn't require the use of any specific authenticator transport and Chrome already supports transports that use UTF-8 and UTF-16LE, so we can't require UTF-8. However, UTF-8 will be common and we need to ensure interoperability. It might be the case that eliminating surrogate code points in UTF-8 is so obvious that it's redundant to specify it here, but it wasn't completely obvious to me. On the other hand, if one writes UTF-8 with surrogate code points to an authenticator, Chrome will not interoperate with it because it considered that to be invalid UTF-8. Thus the desire to highlight that aspect of the UTF-8 spec.

> I don't know how to square replacing "any partial code point at the end with U+FFFD" with the fact that U+FFFD is a 3-byte sequence.

This is platform behaviour in the face of a truncated encoding. Some decoding libraries will replace invalid code point encodings with U+FFFD.

File: index.bs [hunk 1]
```diff
@@ -3072,6 +3072,19 @@ This operation is ignored if it is invoked in an [=authenticator session=] which
...
+
+Additionally, a grapheme cluster may be truncated, <i>altering</i> the resulting character in the mind of the user.
+
+User agents are tasked with handling this to avoid burdening authenticators with understanding character encodings and Unicode character properties. Thus, when dealing with [=authenticators=], user agents SHOULD:
+
+1. Ensure that any strings sent to authenticators are validly encoded. If UTF-8 is used then this includes checking that no code points between U+D800 and U+DFFF (inclusive) are included since those values are not allowed in UTF-8.
+1. Handle the case where strings have been truncated resulting in an invalid encoding. For example, any partial code point at the end may be dropped or replaced with U+FFFD.
```
Comment (equalsJeffH) [inline index.bs]: suggest:
s/ U+FFFD / \<a href='http://unicode.org/cldr/utility/character.jsp?a=FFFD'> U+FFFD \</a> /

File: index.bs [hunk 2]
```diff
@@ -3072,6 +3072,19 @@ This operation is ignored if it is invoked in an [=authenticator session=] which
 or [=authenticatorGetAssertion=] operation currently in progress.
 
 
+## String Handling ## {#sctn-strings}
+
+Authenticators may be required to store arbitrary strings chosen by a [=[RP]=], for example the {{PublicKeyCredentialEntity/name}}, {{PublicKeyCredentialUserEntity/displayName}}, and {{PublicKeyCredentialEntity/icon}} members in a {{PublicKeyCredentialUserEntity}}. Each will have some accommodation for the potentially limited resources available to an [=authenticator=].
+
+If that accommodation is that [=authenticators=] may truncate a value after a specific number of bytes then that MAY be done without reference to any string encoding. For example, an authenticator may truncate a {{PublicKeyCredentialEntity/name}} and leave a partial UTF-8 sequence at the end. If the authenticator is using [[!FIDO-CTAP]] then future messages from the authenticator may contain invalid CBOR since the value is typed as a CBOR string and thus is required to be valid UTF-8.
+
+Additionally, a grapheme cluster may be truncated, <i>altering</i> the resulting character in the mind of the user.
```
[the remark this answers is not in the document]
Comment (agl) [reply index.bs]: Done.

File: index.bs [hunk 3]
```diff
@@ -3072,6 +3072,19 @@ This operation is ignored if it is invoked in an [=authenticator session=] which
...
+
+User agents are tasked with handling this to avoid burdening authenticators with understanding string encodings and Unicode character properties. Thus, when dealing with [=authenticators=], user agents SHOULD:
+
+1. Ensure that any strings sent to authenticators are validly encoded. If UTF-8 is used then this includes checking that no code points between U+D800 and U+DFFF (inclusive) are included since those values are not allowed in UTF-8.
```
[the remark this answers is not in the document]
Comment (agl) [reply index.bs]: This point is discussing the protocol between the platform and the authenticator, which could be anything from Webauthn's perspective. CTAP2 happens to be UTF-8, but Hello is UTF-16LE.

I've added &ldquo;in the authenticator communications protocol&rdquo; to clarify this.

File: index.bs [hunk 4]
```diff
@@ -3072,6 +3072,19 @@ This operation is ignored if it is invoked in an [=authenticator session=] which
 or [=authenticatorGetAssertion=] operation currently in progress.
 
 
+## String Handling ## {#sctn-strings}
+
+Authenticators may be required to store arbitrary strings chosen by a [=[RP]=], for example the {{PublicKeyCredentialEntity/name}}, {{PublicKeyCredentialUserEntity/displayName}}, and {{PublicKeyCredentialEntity/icon}} members in a {{PublicKeyCredentialUserEntity}}. Each will have some accommodation for the potentially limited resources available to an [=authenticator=].
+
+If that accommodation is that [=authenticators=] may truncate a value after a specific number of bytes then that MAY be done without reference to any string encoding. For example, an authenticator may truncate a {{PublicKeyCredentialEntity/name}} and leave a partial UTF-8 sequence at the end. If the authenticator is using [[!FIDO-CTAP]] then future messages from the authenticator may contain invalid CBOR since the value is typed as a CBOR string and thus is required to be valid UTF-8.
+
```
[the remark this answers is not in the document]
Comment (agl) [reply index.bs]: > Won't authnrs likely be required to store arbitrary strings chosen by [=[RP]=]s?

Well, the vast majority of authenticators at the moment are U2F and thus don't store any RP-chosen state. Even CTAP2 authenticators don't have to support discoverable credentials.

> suggest:

Done.

File: index.bs [hunk 5]
```diff
@@ -3072,6 +3072,19 @@ This operation is ignored if it is invoked in an [=authenticator session=] which
 or [=authenticatorGetAssertion=] operation currently in progress.
 
 
+## String Handling ## {#sctn-strings}
+
+Authenticators may be required to store arbitrary strings chosen by a [=[RP]=], for example the {{PublicKeyCredentialEntity/name}}, {{PublicKeyCredentialUserEntity/displayName}}, and {{PublicKeyCredentialEntity/icon}} members in a {{PublicKeyCredentialUserEntity}}. Each will have some accommodation for the potentially limited resources available to an [=authenticator=].
```
[the remark this answers is not in the document]
Comment (agl) [reply index.bs]: For the icon URL, the accommodation is that authenticators may discard it if it's longer than 128 bytes. I.e. the icon case highlights that truncation is not always the fallback.

File: index.bs (hunk 1 above)
Comment (agl) [reply index.bs]: Done.

**Comment by aphillips** [comment]
@agl said:

> It might be the case that eliminating surrogate code points in UTF-8 is so obvious that it's redundant to specify it here, but it wasn't completely obvious to me. On the other hand, if one writes UTF-8 with surrogate code points to an authenticator, Chrome will not interoperate with it because it considered that to be invalid UTF-8. Thus the desire to highlight that aspect of the UTF-8 spec.

I think the disconnect here is: a UTF-16 => UTF-8 transcoder will convert a surrogate pair into a 4-byte UTF-8 character. Anything that encodes the surrogates separately is non-compliant (and in practice these don't exist). The only source for isolated surrogates in UTF-8 would be unpaired surrogates in the data source (there are of course things that can cause this in UTF-16 encoded strings, such as splitting a string arbitrarily...) What I'd way to avoid is people writing lots of machinery to check for this state, when in practice their transcoder and the [[Encoding]] spec have already handled this.

> >    I don't know how to square replacing "any partial code point at the end with U+FFFD" with the fact that U+FFFD is a 3-byte sequence.
>
> This is platform behaviour in the face of a truncated encoding. Some decoding libraries will replace invalid code point encodings with U+FFFD.

No, I understood that to be the case: if reading a UTF-8 (or UTF-16 fwiw) string that has dangling bytes (or other malformed sequences), the dangling bytes are replaced by the transcoder with U+FFFD. However, my comment was basically: I read this as saying to the *generator* (the one doing the truncating) to replace the character with U+FFFD. If one is trying to fit a byte length limit and only has 2 bytes left, you can't put a 3-byte sequence there :-)

File: index.bs [hunk 6]
```diff
@@ -3076,11 +3076,11 @@ or [=authenticatorGetAssertion=] operation currently in progress.
...
 
-If that accommodation is that [=authenticators=] may truncate a value after a specific number of bytes then that MAY be done without reference to any string encoding. For example, an authenticator may truncate a {{PublicKeyCredentialEntity/name}} and leave a partial UTF-8 sequence at the end. If the authenticator is using [[!FIDO-CTAP]] then future messages from the authenticator may contain invalid CBOR since the value is typed as a CBOR string and thus is required to be valid UTF-8.
+If that accommodation is that [=authenticators=] may truncate a value after a specific number of bytes, the authenticator SHOULD take care to do so on character boundaries, although the authenticator MAY truncate without reference to any string encoding. For example, an authenticator could truncate a {{PublicKeyCredentialEntity/name}} and leave a partial UTF-8 sequence at the end. If the authenticator is using [[!FIDO-CTAP]] then future messages from the authenticator may contain invalid CBOR since the value is typed as a CBOR string and thus is required to be valid UTF-8.
```
Comment (aphillips) [inline index.bs]: I'd replace "string encoding" with "character encoding".

**Comment by dwaite** [comment]
The client can detect it was provided a truncated code point in UTF-8 after-the-fact, but won't be able to necessarily detect a truncated grapheme. The grapheme combining rules are also too complex to be static - they can change in later unicode releases, and you could also have fun edge cases like the Taiwanese flag on devices sold in mainland China. 

I believe the RP information is only ever displayed by the authenticator itself, so if it wishes to alter the data (such as truncating it) then it is responsible for doing so appropriately.

I suspect from reading the CTAP spec that client could detect truncation of user metadata by doing a getAssertion following makeCredential, and then potentially do a second makeCredential supplying a better form. But I don't think this would then meet the current webauthn spec, as this would be truncating below the minimum maximum of 64 bytes.

File: index.bs (hunk 5 above)
[the remark this answers is not in the document]
Comment (equalsJeffH) [reply index.bs]: ok.

File: index.bs (hunk 6 above)
Comment (agl) [reply index.bs]: Done in 12763df.

**Comment by agl** [comment]
> What I'd want to avoid is people writing lots of machinery to check for this state, when in practice their transcoder and the [[Encoding]] spec have already handled this.

Ok, thanks for the info. I have dropped this wording.

> The client [&hellip;] won't be able to necessarily detect a truncated grapheme.

That is correct. Hopefully truncation does not happen often and authenticators are more generous with space than strictly required.

> I believe the RP information is only ever displayed by the authenticator itself, so if it wishes to alter the data (such as truncating it) then it is responsible for doing so appropriately.

Authenticators do not have to have displays. If they do not then an account selection dialog is shown by the user-agent if there is more than one credential to choose between. That dialog will contain the RP information as provided (and truncated) by the authenticator.

**Comment by dwaite** [comment]
> > I believe the RP information is only ever displayed by the authenticator itself, so if it wishes to alter the data (such as truncating it) then it is responsible for doing so appropriately.
> 
> Authenticators do not have to have displays. If they do not then an account selection dialog is shown by the user-agent if there is more than one credential to choose between. That dialog will contain the RP information as provided (and truncated) by the authenticator.

This is not my understanding with CTAP based on the January publication - I don't believe there is a case where a PublicKeyCredentialRpEntity is returned to the client.

In any case, an authenticator value that truncates the text to the min-max cannot be further truncated by the client to account for incomplete codepoints/graphemes, since that would put it below the WebAuthn min-max. Such effort would also I believe currently require a second makeCredential with up/uv.

**Comment by agl** [comment]
> This is not my understanding with CTAP based on the January publication.

This PR resulting from having implemented this functionality in Chrome. It reflects what future releases of Chrome and (I'm told) Windows will do. It's already the case that with conformant authenticators on the market that one can use Webauthn to create a discoverable credential with a `name` or `displayName` greater than 64 bytes and then get them to return invalid CBOR to future requests due to string truncation. Platforms have to handle this.

**Comment by agl** [comment]
Based on call of 2019-05-15: will merge this to make it into draft 1. But it's only draft 1 and any issues can still be addressed. But this reflects current implementation practice and does document a subtle problem.

**Pull request merged**